In [97]:
import pandas as pd
import numpy as np
from datetime import datetime

In [98]:
new_dataset = pd.read_csv('/train.csv')

In [99]:
new_dataset.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  New_Price  Price  
0  126.2 bhp    5.0        NaN  12.50  
1   88.7 bhp    5.0  8.61 Lakh   4.50  
2  88.76 bhp    7.0        NaN   6.00  
3  140.8 bhp    5.0        NaN  17.74  
4   63.1 bhp    5.0        NaN   3.50

(a&b)Handle missing values and remove units from the columns

In [100]:
# Check for missing values in each column
missing_values1 = new_dataset.isnull().sum()
print("Missing Values in each column are:")
print(missing_values1)

Missing Values in each column are:
Unnamed: 0              0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 36
Power                  36
Seats                  38
New_Price            5032
Price                   0
dtype: int64


By above observation, Mileage, Engine, Power, Seats & New_Price has null or missing values.
1. Handle Mileage column
The error shows sample('19.67 kmpl') cannot convert to numeric.

Solution: Remove the units.

In [101]:
# Check the missing values
print("Number of Missing Values in Mileage Column before handling are:" ,  new_dataset['Mileage'].isnull().sum())

# Regular Expression is used in str.extract() function to extract substrings
# \d+: Matches one or more digits.
# \.: Matches a literal dot (period).
# \d+: Matches one or more digits after the dot.
# astype(float) function is then used to convert the extracted substrings(str) to float.
new_dataset['Mileage'] = new_dataset['Mileage'].str.extract('(\d+\.\d+|\d+)').astype(float)

# Mileage is reported in kmpl,
# it's reasonable to assume that missing values are likely to be missing at random.
# In such cases, filling with the median is a common strategy.
new_dataset['Mileage'].fillna(new_dataset['Mileage'].median(), inplace=True)

# Check missing values
print(" Number of Missing Values in Mileage Column after handling are:" ,  new_dataset['Mileage'].isnull().sum())

Number of Missing Values in Mileage Column before handling are: 2
 Number of Missing Values in Mileage Column after handling are: 0


2. Handle Engine column
The error shows sample('300 CC') cannot convert to numeric.

Solution: Remove the units.

In [102]:
# Check missing values in the dataset
print("Number of Missing Values in Engine Column after handling are:" ,  new_dataset['Engine'].isnull().sum())
# Remove units
new_dataset['Engine'] = new_dataset['Engine'].str.extract('(\d+\.\d+|\d+)').astype(float)
# Missing values for engine size are often handled by filling the mean value
new_dataset['Engine'].fillna(new_dataset['Engine'].mean(), inplace=True)
# Check missing values in the dataset
print("Number of Missing Values in Engine Column before handling are:" , new_dataset['Engine'].isnull().sum())

Number of Missing Values in Engine Column after handling are: 36
Number of Missing Values in Engine Column before handling are: 0


3. Handle Power Column
The error shows sample('39.09 bhp') cannot convert to numeric.

Solution: Remove the units.

In [103]:
# Check missing values in the dataset
print("Number of Missing Values in Power Column after handling:" ,  new_dataset['Power'].isnull().sum())
# Remove units in the dataset
new_dataset['Power'] = new_dataset['Power'].str.extract('(\d+\.\d+|\d+)').astype(float)
# Missing values for Power are often handled by filling the mean value
new_dataset['Power'].fillna(new_dataset['Power'].mean(), inplace=True)
# Check missing values in the dataset
print("Number of Missing Values in Power Column before handling are:" , new_dataset['Power'].isnull().sum())

Number of Missing Values in Power Column after handling: 36
Number of Missing Values in Power Column before handling are: 0


4. Handle Seats Column

In [104]:
# Check missing values in the dataset
print("Number of Missing Values in Seats Column before handling are:" ,  new_dataset['Seats'].isnull().sum())
# Missing values for seats of the car are often handled by filling the mean value
new_dataset['Seats'].fillna(new_dataset['Seats'].mean(), inplace=True)
# Check missing values in the dataset
print("Number of Missing Values in Seats Column after handling are:" ,  new_dataset['Seats'].isnull().sum())

Number of Missing Values in Seats Column before handling are: 38
Number of Missing Values in Seats Column after handling are: 0


5. Handle New_Price Column

In [105]:
# Drop the New_Price column because it has a high number of missing values
new_dataset.drop('New_Price', axis=1, inplace=True)

View the no null values handled DataFrame.

In [106]:
missing_values = new_dataset.isnull().sum()
missing_values

Unnamed: 0           0
Name                 0
Location             0
Year                 0
Kilometers_Driven    0
Fuel_Type            0
Transmission         0
Owner_Type           0
Mileage              0
Engine               0
Power                0
Seats                0
Price                0
dtype: int64

(c)  Change the categorical variables (“Fuel_Type” and “Transmission”) into numerical one hot encoded value

In [107]:
# Encoding Categorical Variables by using get_dummies function from pandas
new_dataset = pd.get_dummies(new_dataset, columns=['Fuel_Type', 'Transmission'])

# Display the modified dataset
print(new_dataset.head())

   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0              True               False             False   
1             Fals

(d)  Create one more feature and add this column to the dataset.

Creating Years_in_service feature which define the age of the car.

In [108]:
# Calculate current year
current_year = datetime.now().year

# Subtract "Year" column from the current year to calculate the age of the car
new_dataset['Yrs_in_Service'] = current_year - new_dataset['Year']

# Display the modified dataset with the new feature
print(new_dataset.head())

   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0              True               False             False   
1             Fals

Creating Mileage_Category feature to define mileage value by category.

In [109]:
# Define the mileage categories of the cars
mileage_bins = [0, 10, 20, 30, float('inf')]
mileage_labels = ['Very Low', 'Low', 'Medium', 'High']

# Create a new feature 'Mileage_Category' based on the mileage groups
new_dataset['Mileage_Category'] = pd.cut(new_dataset['Mileage'], bins=mileage_bins, labels=mileage_labels, right=False)

# Display the modified dataset with the new feature
print(new_dataset.head())

   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0              True               False             False   
1             Fals

(e) Perform select, filter, rename, mutate, arrange and summarize with group by operations (or their equivalent operations in python) on this dataset.

In [110]:
# Selecting required columns in the dataset
selected_data = new_dataset[['Name', 'Location', 'Year', 'Mileage', 'Price', 'Yrs_in_Service']]
# Displaying the results in the dataset
print("Selected Data:")
print(selected_data.head())

Selected Data:
                               Name    Location  Year  Mileage  Price  \
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015    19.67  12.50   
1                      Honda Jazz V     Chennai  2011    13.00   4.50   
2                 Maruti Ertiga VDI     Chennai  2012    20.77   6.00   
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013    15.20  17.74   
4            Nissan Micra Diesel XV      Jaipur  2013    23.08   3.50   

   Yrs_in_Service  
0               9  
1              13  
2              12  
3              11  
4              11  


In [111]:
# Filtering data based on the Year
filtered_data = new_dataset[new_dataset['Year'] >= 2017]
print("\nFiltered Data (Year >= 2019):")
print(filtered_data.head())


Filtered Data (Year >= 2019):
    Unnamed: 0                                     Name Location  Year  \
8           10                         Maruti Ciaz Zeta    Kochi  2018   
15          17                    Maruti Swift DDiS VDI   Jaipur  2017   
26          28                      Honda WRV i-VTEC VX    Kochi  2018   
34          36            Maruti Vitara Brezza ZDi Plus    Kochi  2018   
35          37  Mahindra KUV 100 mFALCON G80 K6 5str AW    Delhi  2017   

    Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
8               25692      First    21.56  1462.0  103.25    5.0   9.95   
15              25000      First    28.40  1248.0   74.00    5.0   5.99   
26              37430      First    17.50  1199.0   88.70    5.0   9.90   
34              50075      First    24.30  1248.0   88.50    5.0   8.63   
35              52000      First    18.15  1198.0   82.00    5.0   4.85   

    Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
8          

In [112]:
# Renaming columns in the dataset
renamed_data = new_dataset.rename(columns={'Engine': 'EngineDisplacement', 'Power': 'HorsePower'})

print("\nRenamed Columns:")
print(renamed_data.head())


Renamed Columns:
   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  EngineDisplacement  HorsePower  \
0              41000      First    19.67              1582.0      126.20   
1              46000      First    13.00              1199.0       88.70   
2              87000      First    20.77              1248.0       88.76   
3              40670     Second    15.20              1968.0      140.80   
4              86999      First    23.08              1461.0       63.10   

   Seats  Price  Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0    5.0  12.50              True  

In [113]:
# Mutating data (creating new columns)
new_dataset['Price_Category'] = pd.cut(new_dataset['Price'], bins=[0, 5, 10, 15, 20, 25, 30], labels=['<5', '5-10', '10-15', '15-20', '20-25', '25-30'])

print("\nMutated Data (Added Price_Category column):")
print(new_dataset.head())


Mutated Data (Added Price_Category column):
   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0              True               

In [114]:
# Arranging data based on Year in descending order
arranged_data = new_dataset.sort_values(by='Year', ascending=False)

print("\n Arranged Data (data Sorted by Year in descending order):")
print(arranged_data.head())


 Arranged Data (data Sorted by Year in descending order):
      Unnamed: 0                             Name    Location  Year  \
5405        5560        Renault KWID RXT Optional       Kochi  2019   
942          975  Ford Endeavour 2.2 Trend AT 4X2       Kochi  2019   
5533        5690        Maruti Omni 5 Seater BSIV  Coimbatore  2019   
770          796            Mahindra XUV500 W9 AT  Coimbatore  2019   
4267        4399       Maruti Swift Dzire AMT ZDI     Chennai  2019   

      Kilometers_Driven Owner_Type  Mileage  Engine  Power  Seats  Price  \
5405               6568      First    25.17   799.0   53.3    5.0   5.09   
942               11209      First    12.62  2198.0  158.0    7.0  31.15   
5533               4721      First    14.00   796.0   35.0    5.0   4.11   
770               19654      First    14.00  2179.0  155.0    7.0  17.63   
4267              65000      First    26.59  1248.0   74.0    5.0   6.75   

      Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Pet

In [115]:
# Summarizing data using group by operations
summarize_data = new_dataset.groupby('Location')['Price'].mean()

print("\n Summarize Data (Average Price by Location):")
print(summarize_data)


 Summarize Data (Average Price by Location):
Location
Ahmedabad      8.567248
Bangalore     13.482670
Chennai        7.958340
Coimbatore    15.160206
Delhi          9.881944
Hyderabad      9.997423
Jaipur         5.916725
Kochi         11.309109
Kolkata        5.733924
Mumbai         9.592546
Pune           6.951000
Name: Price, dtype: float64


Save raw_data(DataFrame) in New CSV

In [116]:
new_dataset.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0              True               False             False   
1             False               False              True   
2              True               False             False   
3              True               False             False   
4              True               False             False   

   Transmission_Automatic  Transmission_Manual  Yrs_in_Service  \
0                   False                 True               9   
1                   False                 True              13   
2                   False                 True              12   
3                    True                False              11   
4                   False                 True              11   

  Mileage_Category Price_Category  
0              Low          10-15  
1              Low             <5  
2           Medium           5-10  
3              Low          15-20  
4           Medium             <5

In [119]:
#Save the data into cleaned data csv
new_dataset.to_csv("/clean_data.csv")